### DISCUSSION TITLE
What is DIO0 used for

### DISCUSSION THREAD

**Discussion by Cathprotech** [sandeepmistry/arduino-LoRa#453] [Q&A] [answered]
Using the AiThinker RA-02 module, I connected it up to a Wemos D1 Mini (ESP8266). I had D4(IO2) of the ESP connected to DIO0 on the RA-02 module.
I found that using the LoRaReceiverCallback example that DIO0 had to be connected to receive data but if I wanted to upload firmware I had to disconnect to allow upload then re-connect it again to enable recived data.
I then changed to the LoRaSetSpread example and found that now I could receive data without DIO0 connected at all.
Currently I am transmitting a beacon every 5 to 10 minutes and receiving data inbetween.
I changed the defaults in the LoRa.h file for the pinout of my ESP8266 before I started, so that hasn't changed between the two examples.

So, what is DIO0 used for, is it needed at all, if so what for, or am I missing something?
Thanks
Martin

**Comment by IoTThinks** [comment, marked as answer]
DIO0 is used for Continous Receive mode. Check the Receive callback example.
It is also used to detect TXDone.

DIO0 is not required for Single Receive mode.
It is ParsePacket().

**Comment by Erik84750** [comment]
Very good library, thank you very much!!
_"DIO0 is used for Continous Receive mode. Check the Receive callback example."_ : I found that this example allows void loop() to perform any tasks without interfering with radio reception. Is that what you mean?

**Reply by IoTThinks** [reply]
I suggest you search for sx1278 data sheet from Semtech and search for Receive Single Mode and Receive Continous Mode.

In RX continuous mode, the chip keeps listening for packet and signal to DIO0.

For Single mode, the chip only listens ONCE when you run parsePacket()

**Comment by Erik84750** [comment]
I do not think this is defined that way in LoRa.cpp: continuous mode seems to be used when DIO0 is not configured as an interrupt; when in callback-mode, the interrupt INT0 (Arduino Uno) is used by DIO0 to signal an arriving message:

Lines 351 to 403:
[code]
```
`void LoRaClass::flush()
{
}

#ifndef ARDUINO_SAMD_MKRWAN1300
void LoRaClass::onReceive(void(*callback)(int))
{
  _onReceive = callback;

  if (callback) {
    pinMode(_dio0, INPUT);
#ifdef SPI_HAS_NOTUSINGINTERRUPT
    SPI.usingInterrupt(digitalPinToInterrupt(_dio0));
#endif
    attachInterrupt(digitalPinToInterrupt(_dio0), LoRaClass::onDio0Rise, RISING);
  } else {
    detachInterrupt(digitalPinToInterrupt(_dio0));
#ifdef SPI_HAS_NOTUSINGINTERRUPT
    SPI.notUsingInterrupt(digitalPinToInterrupt(_dio0));
#endif
  }
}

void LoRaClass::onTxDone(void(*callback)())
{
  _onTxDone = callback;

  if (callback) {
    pinMode(_dio0, INPUT);
#ifdef SPI_HAS_NOTUSINGINTERRUPT
    SPI.usingInterrupt(digitalPinToInterrupt(_dio0));
#endif
    attachInterrupt(digitalPinToInterrupt(_dio0), LoRaClass::onDio0Rise, RISING);
  } else {
    detachInterrupt(digitalPinToInterrupt(_dio0));
#ifdef SPI_HAS_NOTUSINGINTERRUPT
    SPI.notUsingInterrupt(digitalPinToInterrupt(_dio0));
#endif
  }
}

void LoRaClass::receive(int size)
{

  writeRegister(REG_DIO_MAPPING_1, 0x00); // DIO0 => RXDONE

  if (size > 0) {
    implicitHeaderMode();

    writeRegister(REG_PAYLOAD_LENGTH, size & 0xff);
  } else {
    explicitHeaderMode();
  }

  writeRegister(REG_OP_MODE, MODE_LONG_RANGE_MODE | MODE_RX_CONTINUOUS);
}
#endif

`
```
[/code]

**Comment by IoTThinks** [comment]
Read the data sheet first.

And search this github.
I also asked the same question before.

Some moderators answered this very comprehensive.

**Comment by IoTThinks** [comment]
Read issue 341

https://github.com/sandeepmistry/arduino-LoRa/issues/341
